docs/gen_images.py

```diff
@@ -66,6 +66,7 @@ def generate_filter_image(
     diag: context.ContextDiagram = obj.context_diagram
     diag.filters.clear()
     diag.filters = {filter_name}
+    diag.invalidate_cache()
```

Comment:
Calling `invalidate_cache()` on a diagram from outside like this should never be necessary - if this is in fact needed here, then I would consider it a bug on the implementation side, and this line a crude hack to work around it.

You also added the `invalidate_cache()` call to the `ContextDiagram.FilterSet.add()` method, and with it already having been called from `discard()`, that makes all abstract mutation methods of `MutableSet`. That means that every mutation to the `FilterSet` (even through mixin methods) should now result in an invalidated cache. Can you maybe double-check that this is the case, and if so remove this line (and the one in `generate_styling_image()` below)?